### ISSUE TITLE
Improve the performance of deleting read docs

### ISSUE THREAD

**Issue by garethbowen** [medic/cht-core#6116]
**Describe the performance issue**
Every time a document is deleted sentinel deletes the associated "read" docs from the users databases. This is so the next time the user replicates the little red bubble isn't inflated (no pun intended) by counting docs that no longer exist. To do this we open up every user's user database and delete the doc if it exists. We do this in series so we don't crash sentinel by trying to open too many dbs at once. This makes sentinel very slow to process deletes.

**Describe the improvement you'd like**
Instead of processing the read deletions in the main sequence procession, introduce a new sequence processor which runs periodically (maybe every 30 minutes?). This processor will get changes from its last checked seq till now and for all deletions delete the corresponding read doc. Furthermore it turns out that the slow part is opening the users databases one by one, so instead, batch up all the deletions for that half an hour and delete them from each users database with a bulkdocs call. That way we only have to open each user's database once per interval.

**Describe alternatives you've considered**
We could just run the deletions without blocking the main processing thread which would solve the responsiveness problem but would lose changes if sentinel crashed and would also not batch the deletions so we'd still end up opening the user's dbs too often thus slowing down the whole process.

**To Reproduce**
Steps to record the performance metrics:
1. Make 1000 users
2. Create 10000 docs
3. Delete 5000 docs
4. Create another doc
5. Measure CPU usage, RAM usage, time for all read docs to be deleted, and time for sentinel to process the doc created in 4

**Measurements**
Take measurements before starting development.

**Environment**
- Instance: localhost, recently happened with MUSO
- Browser: N/A
- Client platform: Linux
- App: sentinel
- Version: 3.7.0, master

**Comment by garethbowen** [comment]
@SCdF How close are we to having this ready for AT? I think it's reasonably important so I don't want to delay it.

**Comment by SCdF** [comment]
Not close unfortunately, I will make this the first major ticket I start back working on, but it's also probably quite destabilising so I think leaving it out of 3.9 would be a sensible call

**Comment by garethbowen** [comment]
Ok. Pushing it to 3.10.0.

**Comment by SCdF** [comment]
FYI back on it

**Comment by garethbowen** [comment]
Ready for AT in `6116_async-nonessential-delete-cleanup`

**Comment by SCdF** [comment]
AT:
 - When you delete a document, any read documents in users dbs (that get created to mark that you've viewed a document) and the info doc for that document should get deleted. In this PR we change it so that instead of happening "instantly" (read: when transitions see the change) this now happens as a background job. This job runs on the standard schedule, so every ~5 minutes or on startup So, ensure that both of these things still happen!
 - We also refactored how schedule execution works to be more robust. Ensure that other schedules still execute as they did before.

**Comment by ngaruko** [comment]
LGTM

**Comment by dianabarsan** [comment]
I'm moving this back to "In progress" because it needs a little more work. 
A migration that it's adding needs a small update to generate the a new seq tracker for sentinel - so when Sentinel v.310 starts for the 1st time after an upgrade, it won't start processing all deletes from 0.

**Comment by dianabarsan** [comment]
This change refactors how Sentinel's scheduled tasks are processed. 
Before, they were run in sequence, blocking next task until done. The 5 minute restart interval was calculated _after_ all tasks were finished. 
Now, they should run in parallel, and the scheduler checks, every 5 minutes, if a task is not already running. if not, it would start it, otherwise it would skip it and check it again next interval. 

This is pretty high risk change and should be AT-ed carefully to make sure we don't destabilize sentinel task execution. 

These are the tasks that the scheduler executes:
- due tasks - marking "scheduled" SMS tasks as "pending" so they'd be queued to be sent 
- reminders
- replications of meta dbs
- outbound
- purging
- backgroundCleanup - this is the new task that should handle deleting read docs from meta dbs and infodocs from sentinel

When AT-ing, we should make Sentinel execute all these tasks in parallel, preferably using long-running tasks, to check for potential bugs. So we need a configuration and DB setup that triggers all these actions to run at once, and possibly for a longer period of time (So the 5 minute interval overlaps with execution) and make sure we're never starting the same task twice. 

This means we need thousands of due_tasks, thousands of reminders, thousands of feedback docs to replicate from multiple user dbs to users-meta, and so on.

**Comment by dianabarsan** [comment]
Ah also, when upgrading an instance to this version and start sentinel, make sure the backgroundCleanup doesn't start processing deletes from 0

**Comment by newtewt** [comment]
@dianabarsan  I believe this is ready to close out. For the cases you listed I performed the following tasks. 

due tasks - created lots of messages which found new bug for null due dates. After the fix the messages are getting marked as pending. 

reminders - Lots of reminders generated when loading code and setting up reminders settings. 

replications of meta dbs  - sent 100ks worth of feedbacks from testing menu with online and offline user. Saw it span over 5 minutes cleaning up. Had to alter the cron to process faster*

outbound - pushed outbound for 7k docs all completed fine.

purging - Purged 43k reports and messages no issues. 

backgroundCleanup - generated a bunch of read reports sync and then deleted. All process with the backgroundCleanup. 

Applied the update and saw the Background cleanup batch: 8247 -> 8272 (23) 

Side question. What do the number represent?  ` Background cleanup batch: 8247 -> 8272 (23) `

**Comment by dianabarsan** [comment]
Did AT include this issue: https://github.com/medic/cht-core/issues/6582 (this has been merged into the same branch as this). 

> Side question. What do the number represent? Background cleanup batch: 8247 -> 8272 (23)

This is the new background cleanup task described in the main issue description and what I was referring to in this comment: https://github.com/medic/cht-core/issues/6116#issuecomment-662824377
It should go over all changes in the database, and delete read docs and infodocs for deletes it processes. 

The main transitions feed should now only create the tombstone when it processes the delete.

**Comment by newtewt** [comment]
No, I didn't realize there was a separate issue created for that. I'll work through it now.

**Comment by dianabarsan** [comment]
I created a separate issue just for having a reference and full description of the bug. Merged into the same branch to help AT.

**Comment by dianabarsan** [comment]
Merged into `master`.
